contracts/BalanceTrackerFixedPrice.sol

```diff
@@ -0,0 +1,332 @@
+
+contract BalanceTrackerFixedPrice {
+    event MechPaymentCalculated(address indexed mech, uint256 indexed requestId, uint256 deliveryRate, uint256 rateDiff);
+    event Deposit(address indexed account, address indexed token, uint256 amount);
+    event Withdraw(address indexed account, address indexed token, uint256 amount);
+    event Drained(address indexed token, uint256 collectedFees);
+
+    // Mech marketplace address
+    address public immutable mechMarketplace;
+    // Wrapped native token address
+    address public immutable wrappedNativeToken;
+    // Buy back burner address
+    address public immutable buyBackBurner;
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Map of requester => map of (token => current debit balance)
+    mapping(address => mapping(address => uint256)) public mapRequesterBalances;
+    // Map of mech => map of (token => current debit balance)
+    mapping(address => mapping(address => uint256)) public mapMechBalances;
+    // Map of token => collected fees
+    mapping(address => uint256) public mapCollectedFees;
+    // Map of requestId => token
+    mapping(uint256 => address) public mapRequestIdTokens;
+
+    /// @dev BalanceTrackerFixedPrice constructor.
+    /// @param _mechMarketplace Mech marketplace address.
+    /// @param _wrappedNativeToken Wrapped native token address.
+    /// @param _buyBackBurner Buy back burner address.
+    constructor(address _mechMarketplace, address _wrappedNativeToken, address _buyBackBurner) {
+        // Check for zero address
+        if (_mechMarketplace == address(0) || _wrappedNativeToken == address(0) || _buyBackBurner == address(0)) {
+            revert ZeroAddress();
+        }
+
+        mechMarketplace = _mechMarketplace;
+        wrappedNativeToken = _wrappedNativeToken;
+        buyBackBurner = _buyBackBurner;
+    }
+
+    function _wrap(uint256 amount) internal virtual {
+        IWrappedToken(wrappedNativeToken).deposit{value: amount}();
+    }
+
+    // Check and record delivery rate
+    function checkAndRecordDeliveryRate(
+        address mech,
+        address requester,
+        uint256 requestId,
+        bytes memory paymentData
+    ) external payable {
+        // Check for marketplace access
+        if (msg.sender != mechMarketplace) {
+            revert ManagerOnly(msg.sender, mechMarketplace);
+        }
+
+        // Get mech max delivery rate
+        uint256 maxDeliveryRate = IMech(mech).maxDeliveryRate();
+
+        // Get payment token
+        address token;
+        if (paymentData.length == 32) {
+            // Extract token address
+            token = abi.decode(paymentData, (address));
+            if (token != address(0) && msg.value > 0) {
+                revert NoDepositAllowed(msg.value);
+            }
+        } else if (paymentData.length > 0) {
+            revert InvalidPayloadLength(paymentData.length, 32);
+        }
+
```

Comment:
There's no handling of msg.value below in case of native token transfer

I never approved giving up on one-off payments per mech requests as we currently support. The new system should support the same way mechs worked in the past as well as new types..

you can trivially apply msg.value to the balance logic below when it is positive and matches the native token type for the balancertracker